src/main/java/seedu/address/model/company/UniqueCompanyList.java

```diff
@@ -0,0 +1,159 @@
+     * @param editedCompany Company to replace the other.
+     */
+    public void setCompany(Company target, Company editedCompany) {
+        requireAllNonNull(target, editedCompany);
+
+        int index = internalList.indexOf(target);
+        if (index == -1) {
+            throw new CompanyNotFoundException();
+        }
+
+        if (!target.equals(editedCompany) && contains(editedCompany)) {
+            throw new DuplicateCompanyException();
+        }
+
+        internalList.set(index, editedCompany);
+    }
+
+    /**
+     * Removes the specified company from the list.
+     *
+     * @param toRemove Company to be removed.
+     */
+    public void remove(Company toRemove) {
+        requireNonNull(toRemove);
+        if (!internalList.remove(toRemove)) {
+            throw new CompanyNotFoundException();
+        }
+    }
+
+    /**
+     * Replaces the list with another list in the other <code>UniqueCompanyList</code>.
+     *
+     * @param replacement The replacement <code>UniqueCompanyList</code>.
+     */
+    public void setCompanies(UniqueCompanyList replacement) {
+        requireNonNull(replacement);
+        internalList.setAll(replacement.internalList);
+    }
+
+    /**
+     * Replaces the list with another <code>List</code>.
+     *
+     * @param companies The replacement <code>List</code>.
+     */
+    public void setCompanies(List<Company> companies) {
+        requireAllNonNull(companies);
+        if (!companiesAreUnique(companies)) {
+            throw new DuplicateCompanyException();
+        }
+
+        internalList.setAll(companies);
+    }
+
+    /**
+     * Returns the backing list as an unmodifiable <code>ObservableList</code>.
+     */
+    public ObservableList<Company> asUnmodifiableObservableList() {
+        return internalUnmodifiableList;
+    }
+
+    @Override
+    public Iterator<Company> iterator() {
+        return internalList.iterator();
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        if (other == this) {
+            return true;
+        }
+
+        // instanceof handles nulls
+        if (!(other instanceof UniqueCompanyList)) {
+            return false;
+        }
+
+        UniqueCompanyList otherUniqueCompanyList = (UniqueCompanyList) other;
+        return internalList.equals(otherUniqueCompanyList.internalList);
+    }
+
+    @Override
+    public int hashCode() {
+        return internalList.hashCode();
+    }
+
+    @Override
+    public String toString() {
+        return internalList.toString();
+    }
+
+    /**
+     * Returns true if all of the companies in the list are unique.
+     *
+     * @param companies List to check.
+     * @return true if all companies in the list are unique.
+     */
+    private boolean companiesAreUnique(List<Company> companies) {
+        for (int i = 0; i < companies.size() - 1; i++) {
+            for (int j = i + 1; j < companies.size(); j++) {
+                if (companies.get(i).equals(companies.get(j))) {
```

Comment:
Potentially need to change `Company::equals` to `Company::isSameCompany` so that the `DuplicateCompanyException` is correctly used.

Reply:
changed